### ISSUE TITLE
Add a PodPresetBinding resource

### ISSUE THREAD

**Issue by pmorie** [kubernetes-incubator/service-catalog#994]
In #917, we added a set of alpha fields to the `Binding` resource to express the facet of how a binding manifests into a user's application via PodPreset.  After some thought and feedback from the community, I think it may be better to separate the concerns related to PodPreset into a separate resource.

Reasons to do this:

1.  The coupling between OSB bindings and PodPresets may be inconvenient for users that want to use other mechanisms to manifest bindings into pods
2.  We should expect that PodPresets will be one of many ways that bindings are consumed in the kubernetes ecosystem - it will likely be impossible or extremely costly to implement all of those methods within the `Binding` resource
3.  PodPreset is still in an alpha level; decoupling it from `Binding` is likely to reduce churn in the API for that resource
4.  Error condition handling is probably simpler and easier to understand if the lifecycle management of a PodPreset for a binding is in a separate resource

Sketch of what this would look like:

```go
type Binding struct {
  // pod preset fields deleted
}

type AlphaPodPresetBinding struct {
  BindingRef v1.LocalObjectReference

  PodPresetTemplate *settingsv1alpha1.PodPresetTemplate
}
```

This new resource would be backed by a controller that watched the Binding resource and managed the lifecycle of the a PodPreset accordingly.  Example: when the referenced binding became ready, the controller backing `PodPresetBinding` would create a PodPreset according to that resource.  When the binding was deleted, the PodPreset would be deleted.

**Comment by pmorie** [comment]
I think this should be done before beta and have set the milestone accordingly.

**Comment by duglin** [comment]
+1
However, I think we may want to allow for more than one binding to be referenced.  Then there's only one of these things to modify as the definition of what the "app" is changes. People can always choose to create separate ones if they want but for cases where all bindings for an app are moved in unison I think this would help.

**Comment by pmorie** [comment]
> However, I think we may want to allow for more than one binding to be referenced.

I don't think that should be in scope for now, but nothing prevents you from building such a resource in another API server.

**Comment by nilebox** [comment]
I agree with @duglin:
> However, I think we may want to allow for more than one binding to be referenced.

You can clearly see the benefits of such resource already mentioned by @duglin, and discussed in https://github.com/kubernetes-incubator/service-catalog/issues/969#issuecomment-310559671.

**Comment by MHBauer** [comment]
Argument 1 & 2 are really the same thing, and I buy that. 

I'm not sure I buy the argument of 3, if the field is marked Alpha. Changing at the code level is the same either way, we have to rebase and adjust to fix compilation.

4, maybe.

The next controller we write needs to be really structured about the states and conditions for state changes the objects will go through. 

Multiple bindings as a slice of BindingRefs, or as multiple PodPresetBindings, one for each binding?

I think @duglin wants the first, from his "only one of these things to modify" comment.

**Comment by pmorie** [comment]
> I'm assuming you forgot to add the podpreset template fields to this new resource, right?

The other pod preset fields were what I was indicating by the `// other fields` comment; I will clarify

> Then there's only one of these things to modify as the definition of what the "app" is changes.

I do not think we will be able to have a single uber-resource that encapsulates all of the facets of a binding, some thoughts on that below.

> W/o extra logic like I'm talking about, what's the benefit of this object over just creating the podpreset and pointing to the secret that the binding is going to create?

To be clear: I am thinking of `PodPresetBinding` as one integration among many that people will likely want to build.  Let's distinguish between the benefits of distinguishing the OSB binding for the user-manifestation integration.  When I say manifestation to users, I'm referring to the many ways people could conceivably what to manifest bindings into their pods (or controllers that create pods, or something else).  I think so far I personally know of a few of these:

1.  1:1 relationship between _key_ returned in a binding and env vars
2.  `VCAP_SERVICES` style manifestation where a single variable contains all bindings
3.  I believe @nilebox has some use-cases for things wholly outside the scope of kubernetes
4.  A resource in an external API server that would do specialized injections just for application build pods

If I can think of that many off the top of my head, I think it makes sense to keep the OSB integration distinct and let people build integrations on top of it.  Overall list of benefits I can think of for distinguishing OSB bindings from how they manifest to users:

1.  Easier to distinguish the status of the OSB binding (which is all the broker cares about, and important to get exactly right) from the status of how the binding is manifested into apps
2.  Reduces coupling to PodPreset; provides a good example for people to implement sibling-integrations (like `VCAP_SERVICES` support, `$YOUR_PET_OTHER_THING_HERE`) on top of the OSB primitives
3.  Mechanics of `Binding` resource are much simpler to understand without knowing about all additional user-manifestation integrations
4.  The additional resource manages the lifecycle of user-manifestation integration - it doesn't do its work until the binding becomes ready, and hooks into the lifecycle of the binding so that it is cleaned up by GC as part of handling unbind
5.  The different examples I listed above are very disparate; I think if we try to make a single Binding thing that has multiple modes, that it will not scale from a maintanance or implementation standpoint

So, with all that said, the specific benefit for `PodPresetBinding` is that it allows you to build a simple, per-OSB-binding integration using `PodPreset` in whatever configuration you desire, that fits into the lifecycle of the OSB binding.

**Comment by pmorie** [comment]
@nilebox

> I agree with @duglin:
>
> >    However, I think we may want to allow for more than one binding to be referenced.
>
> You can clearly see the benefits of such resource already mentioned by @duglin, and discussed in #969 (comment).

I am really saying that you can build whatever flavor of integration that you want.  I personally think that the scope that has been discussed from the start of this group is satisfied by just what I described as `PodPresetBinding`.   The multiple-binding dependencies to single PodPreset might be a valid use-case, but in my opinion it is probably complex to get all of the different interactions correct.  It hasn't been in the scope we've discussed yet, so I suggest that it either wait until after beta, or that someone prototype it in their own API server and demo it to the SIG once they feel like it's ready to share.

I'd be happy to publish a golang client from this repo to aid such endeavors.

**Comment by kibbles-n-bytes** [comment]
+1
It's come to Google's attention as well that there's a clear desire to allow for flexible resource consumption. Applications that are built on Kubernetes but integrate with Hashicorp's Vault for credentials consumption is an example we've come across when discussing integrating Service Catalog in existing environments. Having an easily extensible mechanism here so that consumers can dictate how the credentials will be injected I think will go a long way to both clarify responsibilities and simplify adoption.

**Comment by nilebox** [comment]
> Applications that are built on Kubernetes but integrate with Hashicorp's Vault for credentials consumption 

That's a good example we are interested in too. Actually it's one of the reasons I am so desperate about using Secrets for storing credentials (see https://github.com/kubernetes-incubator/service-catalog/issues/621 and https://github.com/kubernetes-incubator/service-catalog/pull/957), since Secrets roadmap has plans to support external secret stores (Vault, Amazon KMS etc), so solving your example would be just a matter of configuring Secret controller

**Comment by kibbles-n-bytes** [comment]
As this proposal is suggesting decoupling injection from the core responsibilities of the Service Catalog, thoughts on having this injection logic live in a separate pod?

This would allow for:

1. Clients to add custom injection policies without having to fork the service catalog
2. Changes to the credential injection logic would not require recompiling/redeploying the catalog
3. Flexibility to use whatever programming language and persistence layer desired

The endpoint to this pod could be specified via a CLI flag or an `InjectionPlugin` resource.

**Comment by nilebox** [comment]
Well the whole point about separate `PodPresetBinding` is just that this is an **optional** (but default) extension on top of Bindings.

And if you want to have your own implementation on top of Bindings, you just don't use `PodPresetBinding` and define your own `MyOwnBinding` with corresponding controller in a separate Pod.

That's exactly what we do: see https://github.com/atlassian/smith/tree/master/examples/service_catalog (it was presented to SIG Service Catalog on 19th June, you can watch a recording if interested).

**Comment by duglin** [comment]
Couple of things come to mind:
1 - I see no value in this extra resource that sits between podpreset and bindings - mainly because I can get the exact same results by just creating a podpreset directly. And, in fact, as a user I'm not sure why I would choose to use a pseudo-podpreset (which is a subset of the real thing).

2 - paul said "The multiple-binding dependencies to single PodPreset might be a valid use-case, but in my opinion it is probably complex to get all of the different interactions correct. It hasn't been in the scope we've discussed yet," - is not true. Let me rephrase the requirement... we need to support multiple services per app.  This means multiple bindings per app. This mean multiple bindings per podpreset if there's one podpreset per app.  Now, perhaps there's more than one podpreset per app, but I think that's kind of at the heart of this discussion - either way, we need to support multiple services/bindings.

Maybe because its late :-)  but I might have convinced myself of a couple of things:
A - If we do not try to include the definition of an app (e.g. the labelSelector) in a Binding, then Bindings should be called something like ServiceCredentialsRequest - meaning they exist just for the purpose of getting a new set of credentials for a service instance and storing the results in a secret. Nothing more. I think the word "Binding" is causing us trouble because we were trying to be too CF-ish - and clearly we can't be - and we can't really "bind" anything at this point in our workflow.  (Although, I want to find a better word than "Credentials" since there are other types of Bindings that we may support in the future - unless we decide to create new ServiceXxxRequest resources for those - which might be ok- dunno).

B - Once we have a secret, we're back into normal Kube-land (out of OSB-land) and the user can do whatever they want. They can hardcode their pod specs to use the secrets or they can use podpresets to auto-inject them. Its up to them. As much as I really wanted to make things easier on people by merging the definition of an app into the binding, the flexibility of Kube just doesn't fit well with that idea. 

The biggest problem I see with this idea is that we need to decide what to use for "app_id" in the broker.CreateBinding() request.  We need to pick something that is natural for people to think about changing when they're moving between apps (whatever their definition of an app is). Perhaps just asking them to provide an "app_name" will work - then they can decide on their own when to use a different value based on their mental model of what an app is and how this secret will be used.  Of course "app" might be hard to reason about, so maybe "scope" ? Dunno.

And finally, if we expect people to create their own resources and controllers to solve what I consider basic usecases then we're letting our users down - and multiple services per app is going to be very common/basic.  

We really need a whiteboard and f2f time :-)

**Comment by duglin** [comment]
In prep for today's design call I wrote down my thoughts on this and end with a proposal:
https://gist.github.com/duglin/6ecaf1c079ed6896a613f82a060e823e

I'll add this text directly into the issue later once I'm sure I'm not tweaking it.

**Comment by duglin** [comment]
On today's design call there was general agreement that we should not embed a PodPreset (or any variant of a PodPreset) into the Binding resource. There were many reasons but the main ones seemed to focus on not promoting, or favoring, one particular pattern for injecting the secret into some other resource.  We decided to let this decision be considered by the group for a few days to see if there are any objections.  If not, we will then work on the next steps... which includes deciding what, if any, additional resources we should consider adding to the model.

**Comment by nilebox** [comment]
Update: the support for embedded PodPreset template was removed from the Binding resource, see https://github.com/kubernetes-incubator/service-catalog/pull/1030

**Comment by duglin** [comment]
changed title to more accurately describe this based on recent changes in the project

**Comment by jberkhahn** [comment]
closing this issue in favor of #2638
